Add Check Support for Java 21 Pattern Matching for Switch Syntax: UnnecessaryParentheses

child of #14961 
I have read check documentation: https://checkstyle.org/checks/coding/unnecessaryparentheses.html#UnnecessaryParentheses
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```
➜  src cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="UnnecessaryParentheses">
        </module>
    </module>
</module>
➜  src cat Test.java 
public class Test {
    record Point(boolean x, boolean y, boolean z) { }

    public void test() {
        Point p = new Point(true, true, true);
        boolean a = true, b = true;
        boolean c = false, d = false;
        if ((a && b) || c) { // violation
        }

        switch (p) {
            case Point(boolean x, boolean y, boolean z)
                    when (x && y) || z -> System.out.println(); // violation
            case Point(boolean x, boolean y, boolean z) -> System.out.println("error");
        }
    }
}
➜  src javac --enable-preview --release 21 Test.java                        
➜  src java -jar checkstyle-10.18.1-SNAPSHOT-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:8:13: Unnecessary parentheses around expression. [UnnecessaryParentheses]
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:13:26: Unnecessary parentheses around expression. [UnnecessaryParentheses]
Audit done.
Checkstyle ends with 2 errors.

```

---

**Describe what you expect in detail.**

This check behaves as I would expect after adding the `EXPR` node as parent. Note that I am using the latest snapshot for this example.
---

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/UnnecessaryParenthesesCheckTest.java
@@ -342,4 +342,19 @@ public class UnnecessaryParenthesesCheckTest extends AbstractModuleTestSupport {
         verifyWithInlineConfigParser(
                 getPath("InputUnnecessaryParenthesesOperator3.java"), expected);
     }
+
+    @Test
+    public void testWhenExpressions() throws Exception {
+        final String[] expected = {
+            "22:33: " + getCheckMessage(MSG_EXPR),
+            "24:32: " + getCheckMessage(MSG_EXPR),
+            "28:26: " + getCheckMessage(MSG_EXPR),
+            "31:31: " + getCheckMessage(MSG_EXPR),
+            "31:32: " + getCheckMessage(MSG_EXPR),
+            "37:44: " + getCheckMessage(MSG_EXPR),
+            "40:44: " + getCheckMessage(MSG_EXPR),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputUnnecessaryParenthesesWhenExpressions.java"), expected);
+    }
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesWhenExpressions.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/unnecessaryparentheses/InputUnnecessaryParenthesesWhenExpressions.java
@@ -1,0 +1,59 @@
+/*
+UnnecessaryParentheses
+tokens = (default)EXPR, IDENT, NUM_DOUBLE, NUM_FLOAT, NUM_INT, NUM_LONG, \
+         STRING_LITERAL, LITERAL_NULL, LITERAL_FALSE, LITERAL_TRUE, ASSIGN, \
+         BAND_ASSIGN, BOR_ASSIGN, BSR_ASSIGN, BXOR_ASSIGN, DIV_ASSIGN, \
+         MINUS_ASSIGN, MOD_ASSIGN, PLUS_ASSIGN, SL_ASSIGN, SR_ASSIGN, STAR_ASSIGN, \
+         LAMBDA, TEXT_BLOCK_LITERAL_BEGIN, LAND, LITERAL_INSTANCEOF, GT, LT, GE, \
+         LE, EQUAL, NOT_EQUAL, UNARY_MINUS, UNARY_PLUS, INC, DEC, LNOT, BNOT, \
+         POST_INC, POST_DEC
+
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.coding.unnecessaryparentheses;
+
+public class InputUnnecessaryParenthesesWhenExpressions {
+
+    void test(Object o) {
+
+        switch (o) {
+            case Integer i when (i > 0) -> {}  // violation 'Unnecessary paren.* around expression'
+
+            case String s when (!s.isEmpty()) -> {}
+            // violation above 'Unnecessary paren.* around expression'
+
+            case Point(boolean x, boolean y, boolean z)
+                    when (x && y) || z -> System.out.println();
+            // violation above 'Unnecessary paren.* around expression'
+
+            case Point p when ((p.x && p.y) || p.z) -> System.out.println();
+            // 2 violations above:
+            //                   'Unnecessary paren.* around expression'
+            //                   'Unnecessary paren.* around expression'
+
+            // until https://github.com/checkstyle/checkstyle/issues/15317
+            case Point2(int x, int y) when ((x >= 0 && y >= 0)) -> {}
+            // violation above 'Unnecessary paren.* around expression'
+
+            case Point2(int x, int y) when (x >= 5 && y >= 5) -> {}
+            // violation above 'Unnecessary paren.* around expression'
+
+            default -> {}
+        }
+
+        switch (o) {
+            case Integer i when i > 0 -> {}
+            case String s when !s.isEmpty() -> {}
+            case Point(boolean x, boolean y, boolean z) when x && y || z -> System.out.println();
+            case Point p when p.x && p.y || p.z -> System.out.println();
+            case Point2(int x, int y) when x >= 0 && y >= 0 -> {}
+            case Point2(int x, int y) when x >= 5 && y >= 5 -> {}
+            default -> {}
+        }
+    }
+
+    record Point(boolean x, boolean y, boolean z) { }
+    record Point2(int x, int y) { }
+}

